### PR TITLE
Implement KubeObjectStatusText extension API

### LINKED ISSUE #1118: Allow to register text to status column of resource list
Extensions should be able to render text (optionally with color) to status column of resource list.

### PR DESCRIPTION
**Pull request by nevalla**
Fixes #1118 

**Usage** 
```typescript
// renderer.ts
export default class PodStatusTextExtension extends LensRendererExtension {
 KubeObjectStatusTexts = [
    {
      kind: "Pod",
      apiVersions: ["v1"],
      resolve: (pod: K8sApi.Pod) => resolve(pod)
    }
  ]
}

// pod-status-resolver.tsx
import { K8sApi } from "@k8slens/extensions";

export function resolve(pod: K8sApi.Pod): K8sApi.KubeObjectStatus {
  if (pod.getName().startsWith("nginx-deployment") {
    return {
      level:  K8sApi.KubeObjectStatusLevel.INFO,
      text: "Image scanned",
    }
  }

  return null
}
```

image

Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

### CODE REVIEW

**Comment by jakolehm** [comment]
why not just:
```typescript
export interface ResourceStatusRegistration {
  kind: string;
  apiVersions: string[];
  resolver: (object: KubeObject) => ResourceStatus | null;
}
```

where `ResourceStatus` is an interface with text & color?

**Comment by nevalla** [comment]
> why not just:
> 
> ```ts
> export interface ResourceStatusRegistration {
>   kind: string;
>   apiVersions: string[];
>   resolver: (object: KubeObject) => ResourceStatus | null;
> }
> ```
> 
> where `ResourceStatus` is an interface with text & color?

Much better...will change it.

**Comment by nevalla** [comment]
Now all resource lists are using new `KubeResourceStatusIcon` component instead of `KubeEventIcon`. New component will display multiple status messages (not only from events) in a tooltip (see the image in the description).

File: src/extensions/registries/index.ts [hunk 1]
```diff
@@ -7,3 +7,4 @@ export * from "./status-bar-registry"
 export * from "./kube-object-detail-registry";
 export * from "./kube-object-menu-registry";
 export * from "./cluster-feature-registry"
+export * from "./resource-status-registry"
```
Comment (jakolehm) [inline src/extensions/registries/index.ts]: This should be named similar to other `kube-object-*` registries, maybe `kube-object-status-registry`?

File: src/renderer/components/kube-resource-status-icon/kube-resource-status-icon.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import { KubeObject } from "../../api/kube-object";
+import { cssNames, formatDuration } from "../../utils";
+import { ResourceStatusRegistration, ResourceStatus, ResourceStatusLevel, resourceStatusRegistry } from "../../../extensions/registries/resource-status-registry"
+import { computed } from "mobx";
+
+interface Props {
+  object: KubeObject;
+}
+
+export class KubeResourceStatusIcon extends React.Component<Props> {
```
Comment (jakolehm) [inline src/renderer/components/kube-resource-status-icon/kube-resource-status-icon.tsx]: `KubeObjectStatusIcon`

**Review by jakolehm — CHANGES_REQUESTED** [review]

File: extensions/event-resource-status/package.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  "version": "0.1.0",
+  "description": "Adds resource status from events",
+  "renderer": "dist/renderer.js",
+  "lens": {
+    "metadata": {},
+    "styles": []
+  },
+  "scripts": {
+    "build": "webpack --config webpack.config.js",
+    "dev": "npm run build --watch"
```
Comment (Nokel81) [inline extensions/event-resource-status/package.json]: To comply with #1201 please add a `"test"` script.

File: extensions/event-resource-status/Makefile [hunk 4]
```diff
@@ -0,0 +1,5 @@
+install-deps:
+	npm install
+
+build: install-deps
+	npm run build
```
Comment (Nokel81) [inline extensions/event-resource-status/Makefile]: To comply with #1201 please add a `"test"` target.

File: src/extensions/registries/index.ts (hunk 1 above)
Comment (nevalla) [reply src/extensions/registries/index.ts]: Renamed

File: src/renderer/components/kube-resource-status-icon/kube-resource-status-icon.tsx (hunk 2 above)
Comment (nevalla) [reply src/renderer/components/kube-resource-status-icon/kube-resource-status-icon.tsx]: Done

File: src/extensions/renderer-api/index.ts [hunk 5]
```diff
@@ -4,9 +4,11 @@
 import * as Component from "./components"
 import * as K8sApi from "./k8s-api"
 import * as Navigation from "./navigation"
+import * as KubeObjectStatus from "./kube-object-status"
 
 export {
   Component,
   K8sApi,
   Navigation,
+  KubeObjectStatus
```
Comment (jakolehm) [inline src/extensions/renderer-api/index.ts]: Do we really need separate namespace for this? 🤔

File: src/extensions/renderer-api/k8s-api.ts [hunk 6]
```diff
@@ -2,6 +2,7 @@ export { isAllowedResource } from "../../common/rbac"
 export { apiManager } from "../../renderer/api/api-manager";
 export { KubeObjectStore } from "../../renderer/kube-object.store"
 export { KubeApi, forCluster, IKubeApiCluster } from "../../renderer/api/kube-api";
+export type { EventStore } from "../../renderer/components/+events/event.store"
```
Comment (jakolehm) [inline src/extensions/renderer-api/k8s-api.ts:5]: I think we need a separate issue about missing store types?

File: src/extensions/renderer-api/index.ts (hunk 5 above)
Comment (nevalla) [reply src/extensions/renderer-api/index.ts]: I would rather not to add new namespace, but it does not fit very well in to any existing namespaces. But I'm open for suggestions.

File: extensions/event-resource-status/package.json (hunk 3 above)
Comment (nevalla) [reply extensions/event-resource-status/package.json]: Done

File: extensions/event-resource-status/Makefile (hunk 4 above)
Comment (nevalla) [reply extensions/event-resource-status/Makefile]: Done

File: src/extensions/renderer-api/index.ts (hunk 5 above)
Comment (jakolehm) [reply src/extensions/renderer-api/index.ts]: Everything else related to "KubeObject" is under `K8sApi`, would that make sense here?

Comment (nevalla) [reply src/extensions/renderer-api/index.ts]: Yes. Moved it under `K8sApi`.

File: src/extensions/renderer-api/kube-object-status.ts [hunk 7]
```diff
@@ -0,0 +1 @@
+export { KubeObjectStatus as Status, KubeObjectStatusLevel as Level } from "../registries/kube-object-status-registry"
```
Comment (jakolehm) [inline src/extensions/renderer-api/kube-object-status.ts]: Does not look right? We should not use generic `Status`/`Level` if these are exposed within `K8sApi` namespace.

File: src/extensions/registries/index.ts [hunk 8]
```diff
@@ -7,3 +7,4 @@ export * from "./status-bar-registry"
 export * from "./kube-object-detail-registry";
 export * from "./kube-object-menu-registry";
 export * from "./cluster-feature-registry"
+export * from "./kube-object-status-registry"
```
Comment (jakolehm) [inline src/extensions/registries/index.ts:10]: Isn't this exporting all `KubeObjectStatus` stuff? Is that intentional?

File: src/extensions/renderer-api/kube-object-status.ts (hunk 7 above)
Comment (nevalla) [reply src/extensions/renderer-api/kube-object-status.ts]: Yes, this is still wrong here. Will fix

File: src/extensions/registries/index.ts (hunk 8 above)
Comment (nevalla) [reply src/extensions/registries/index.ts:10]: I moved `KubeObjectStatus` types to `K8sApi`. Feels more natural to have them there.

File: src/extensions/renderer-api/kube-object-status.ts (hunk 7 above)
Comment (nevalla) [reply src/extensions/renderer-api/kube-object-status.ts]: Done

File: extensions/kube-object-event-status/renderer.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  kubeObjectStatusTexts = [
+    {
+      kind: "Pod",
+      apiVersions: ["v1"],
+      resolve: (object: K8sApi.Pod) => resolveStatusForPods(object)
+    },
+    {
+      kind: "ReplicaSet",
+      apiVersions: ["v1"],
+      resolve: (object: K8sApi.KubeObject) => resolveStatus(object)
```
Comment (jakolehm) [inline extensions/kube-object-event-status/renderer.tsx]: Why some of these use specific types and other use generic `KubeObject`?

File: src/extensions/renderer-api/k8s-api.ts [hunk 10]
```diff
@@ -29,3 +30,15 @@ export { Role, roleApi } from "../../renderer/api/endpoints";
 export { RoleBinding, roleBindingApi } from "../../renderer/api/endpoints";
 export { ClusterRole, clusterRoleApi } from "../../renderer/api/endpoints";
 export { ClusterRoleBinding, clusterRoleBindingApi } from "../../renderer/api/endpoints";
+
+export type KubeObjectStatus = {
```
Comment (jakolehm) [inline src/extensions/renderer-api/k8s-api.ts]: Does not feel right to have them here, everything else is exported from somewhere outside.

File: extensions/kube-object-event-status/renderer.tsx (hunk 9 above)
Comment (nevalla) [reply extensions/kube-object-event-status/renderer.tsx]: `resolveStatusFromPods` and `resolveStatusFromCronJobs` requires specific type. I've changed this now to use always specific type to be more consistent.

File: src/extensions/renderer-api/k8s-api.ts (hunk 10 above)
Comment (nevalla) [reply src/extensions/renderer-api/k8s-api.ts]: Moved these types to separate file where those are exported from now

**Comment by nevalla** [comment]
Tests are now again green. @jakolehm PTAL

**Review by jakolehm — APPROVED** [review]
LGTM

**Comment by jakolehm** [comment]
Just realised this PR was missing docs.

**Pull request merged**